### ISSUE TITLE
ClassRule/TestRule version of Mockito's JUnitRule

### ISSUE THREAD

**Issue by magneticflux-** [mockito/mockito#997]
As it currently stands, `JUnitRule` is only usable through the `@Rule` annotation, not the more powerful `@ClassRule` annotation. This prevents it from solving several issues, such as:
- Using the `@Mock` annotation with static fields
- Using the `@Mock` annotation to initialize before the JUnit Theories runner acquires `@DataPoint` annotated fields

I propose a version of Mockito's JUnitRule that implements `TestRule` rather than `MethodRule`. Requirements for this change: 
- A new `MockitoAnnotations.initMocks(target)` method that takes a class rather than an instance and only initializes _static_ fields (`setField(null,value)`)
- A new version of `TestFinishedEvent` that does not contain an instance of the target (The instance is actually never used - only the class name. It could be replaced with a String similar to `getTestMethodName()`)
- A constructor that takes a target `Object` as a parameter to allow the rule to still initialize non-static fields

All of this, of course, hinges on several assumptions. It seems like the functionality provided by this rule should be divided into a non-static component (validation, mocking non-static `@Mock` fields) that is run after every test as an `@Rule` and a static component (no validation, mocking static `@Mock` fields). This would allow Mockito to play nicely with certain JUnit runners such as the Theories runner (See junit-team/junit4/issues/1433) while still providing correct error handling when needed.

Useful facts for reference:
- The Theories runner copies field data _after_ `@BeforeClass` and `@ClassRule` execute, but _before_ `@Before` and `@Rule` execute
- The Theories runner requires `@DataPoint` fields to be static
- The `MockitoAnnotations.initMocks(target)` method is run before every test in the class since it is in a `@Rule`
- An IndependentAnnotationEngine can be used to initialize classes with _only_ static `@Mock` fields if the `testInstance` is `null`
- 

[Here](https://gist.github.com/magneticflux-/4c3478e21540b3f3652ec041752a36a3) is an example of such as class that works for _only_ static fields. It will throw an exception if a non-static `@Mock` field is found because it uses the existing annotation engine.

TL;DR:
Mockito's MethodRule initializes `@Mock` fields too late - the Theories JUnit runner copies the field data before any `@Test`/`@Theory` methods are run, resulting in `null` being passed as a parameter and mocks never being found. Having a `TestRule` (or even just a static method in `MockitoAnnotations`) for initializing static `@Mock` fields would allow these fields to have values when the Theories runner looks for them.

**Comment by szczepiq** [comment]
Thank you for submitting this ticket and a very comprehensive description + research! One of the core devs on the team (perhaps myself :) need to dig into the suggested design.

My quick thoughts:

 - it is unlikely that we will change the interface of the rule in 2.x. It would be an incompatible change that we can do only in a major version
 - in Mockito v3 we would need to decide if changing the rule interface is worth it. JUnit Theories are not really widely used (you are first to suggest the change, although JUnit rule has been around for years). Also, users can work around by not using mockito rule when theories are in use. Users can also create a rule that works for them.
 - creating a new kind of rule and keep it along the existing rule will fragment Mockito API and may not be a good idea

> Having a TestRule (or even just a static method in MockitoAnnotations) for initializing static @Mock fields would allow these fields to have values when the Theories runner looks for them.

MockitoAnnotations has a static method "initMocks"

Can you put together a TestRule implementation that fixes your problem (including a test that demonstrates how it works) and submit a PR? This would be helpful for us to decide on the next steps.

Thanks again for the report! I suspect that we can come up with a solution that can solve your use case cleanly.

**Comment by magneticflux-** [comment]
Thanks for the response! To clarify some things:
- My current solution of initializing the fields using a static `@BeforeClass` method has been working fine, and I completely understand why you're hesitant to support a little-used/experimental feature such as Theories. However, I _also_ feel that extending support would force the current JUnit integration to become more robust.
- The method `initMocks` that I referred to is indeed a method in the `MockitoAnnotations` class, but it requires an instance of the class and it tries to initialize both static _and_ non-static fields of that instance. I would like a companion method that takes only a class, and tries to initialize only static fields.

After some more speculation, I feel that splitting up the existing rule into two rules is better in the long run. This, of course, would be a breaking change slated for Mockito 3.x.

I'm also under the assumption that re-initializing the mocks after each test is _not_ intended. Since a `@Rule` is run before each method, the `initMocks(target)` method is repeatedly called, creating new mocks for every field. For reference, the actual JUnit runner that this `@Rule` aims to replace only runs the `initMocks(target)` method _once_, as does putting `initMocks(target)` in a `@Before` method. Given that information, I believe that the correct behavior for the `@Rule` would be to run once, before the entire suite of test methods.

Now that that's out of the way, these are the problems I see with the current Rule:
- We need a way to initialize static fields **once** with `@Mock`
- We need a way to initialize instance fields **once** with `@Mock`
- We need a way to call the Mockito verification methods after _each_ test (to immediately detect improper usage, per documentation)

I claim that these three items _can not_ and _should not_ be dealt with in the same class, because doing so would violate several principles of object-oriented design and serve to make the API confusing to use.

These are, respectively, the solutions I see for each problem:
- Create a `TestRule` extension that then iterates over static fields using a new `AnnotationEngine` subclass. It has all the information right there in the Statement and Description provided in the `apply` method. This would be intended to be used with the `@ClassRule` annotation.
- I currently do not know of any Rule-based solution to running a non-static method _exactly once_. Using the `@Before` annotation on a method accomplishes this, but is not a rule :(.
- Create a `TestRule` extension that merely calls `Statement.execute()` and `Mockito.validateMockitoUsage()`. This would be intended to be used with the `@Rule` annotation.

**Comment by magneticflux-** [comment]
After only a few more minutes of research, I seem to have invalidated several of my assumptions... I realize now that JUnit instantiates a new object of the test class for _each_ test method it will run so that it is not possible to share data from one test to another through instance fields. I will need to go back over my ideas and figure out how they need to be changed, but for now I assume that everything above is wrong.
